### PR TITLE
External consent page redirection when external consent management is enabled

### PR DESCRIPTION
**Pull request by sandushi**
### Proposed changes in this pull request

- Add the changes for the external consent page redirection
    - Check whether the external consent management enabled for the application
    - Get the configured external consent page url
    - Redirect to the configured external consent page after user is authenticated. 

### Follow up actions

[List any possible follow-up actions here; for instance, testing data
migrations, software that we need to install on staging and production
environments.]

-

#### General

- [ ] **Is this PR explained thoroughly?** All code changes must be accounted for in the PR description.
- [ ] **Is the PR labeled correctly?**

#### Functionality

- [ ] **Are all requirements met?** Compare implemented functionality with the requirements specification.
- [ ] **Does the UI work as expected?** There should be no Javascript errors in the console; all resources should load. There should be no unexpected errors. Deliberately try to break the feature to find out if there are corner cases that are not handled.

#### Code

- [ ] **Do you fully understand the introduced changes to the code?** If not ask for clarification, it might uncover ways to solve a problem in a more elegant and efficient way.
- [ ] **Does the PR introduce any inefficient database requests?** Use the debug server to check for duplicate requests.
- [ ] **Are all necessary strings marked for translation?** All strings that are exposed to users via the UI must be [marked for translation](https://docs.djangoproject.com/en/1.10/topics/i18n/translation/).

#### Tests

- [ ] **Are there sufficient test cases?** Ensure that all components are tested individually; models, forms, and serializers should be tested in isolation even if a test for a view covers these components.
- [ ] **If this is a bug fix, are tests for the issue in place?**  There must be a test case for the bug to ensure the issue won’t regress. Make sure that the tests break without the new code to fix the issue.
- [ ] **If this is a new feature or a significant change to an existing feature?** has the manual testing spreadsheet been updated with instructions for manual testing?

#### Security

- [ ] **Confirm this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.**
- [ ] **Are all UI and API inputs run through forms or serializers?**
- [ ] **Are all external inputs validated and sanitized appropriately?**
- [ ] **Does all branching logic have a default case?**
- [ ] **Does this solution handle outliers and edge cases gracefully?**
- [ ] **Are all external communications secured and restricted to SSL?**

#### Documentation

- [ ] **Are changes to the UI documented in the platform docs?** If this PR introduces new platform site functionality or changes existing ones, the changes should be documented.
- [ ] **Are changes to the API documented in the API docs?** If this PR introduces new API functionality or changes existing ones, the changes must be documented.
- [ ] **Are reusable components documented?** If this PR introduces components that are relevant to other developers (for instance a mixin for a view or a generic form) they should be documented in the Wiki.

### CODE REVIEW

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java [hunk 1]
```diff
@@ -828,11 +832,36 @@ public static String getUserConsentURL(OAuth2Parameters params, String loggedInU
...
+        ServiceProvider sp = null;
+        if (params != null) {
+            sp = OAuth2Util.getServiceProvider(params.getClientId());
+        }
+        return sp;
+    }
+
+    private static boolean isExternalConsentManagementEnabledForSP(ServiceProvider sp) {
+        return FrameworkUtils.isExternalConsentManagementEnabledForSP(sp);
+    }
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java:852]: why do we need this method?
Can't we inline with static imports if you are doing this just for readability

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java [hunk 2]
```diff
@@ -828,11 +832,36 @@ public static String getUserConsentURL(OAuth2Parameters params, String loggedInU
...
             throw new OAuthSystemException("Error while encoding the url", e);
+        } catch (IdentityOAuth2Exception e) {
+            throw new OAuthSystemException("Error retrieve Service Provider for clientId:" + params.getClientId() , e);
         }
 
         return consentPageUrl;
     }
 
+    private static ServiceProvider getServiceProvider(OAuth2Parameters params) throws IdentityOAuth2Exception {
+        ServiceProvider sp = null;
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java:843]: Follow consistent formatting.

Either leave a new line after each method declaration or don't leave at all. 
Follow the same convention everywhere to be consistent.

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java [hunk 3]
```diff
@@ -828,11 +832,36 @@ public static String getUserConsentURL(OAuth2Parameters params, String loggedInU
...
+        return sp;
+    }
+
+    private static boolean isExternalConsentManagementEnabledForSP(ServiceProvider sp) {
+        return FrameworkUtils.isExternalConsentManagementEnabledForSP(sp);
+    }
+
+    private static String getExternalConsentUrlForSP(ServiceProvider sp) throws OAuthSystemException {
+
+        String externalConsentUrl = FrameworkUtils.getExternalConsentUrlForSP(sp);
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java:856]: Why do we need a util method for this?
Can't we get this using the ServiceProvider object?

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java (hunk 1 above)
Comment (sandushi) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java]: Resolved by https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/2030/commits/43c3a582c5d41fc44d849e3f3af1be6550d893ee

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java (hunk 3 above)
Comment (sandushi) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java]: Resolved by https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/2030/commits/43c3a582c5d41fc44d849e3f3af1be6550d893ee

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java (hunk 2 above)
Comment (sandushi) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java:845]: resolved by https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/2030/commits/43c3a582c5d41fc44d849e3f3af1be6550d893ee

**Review by Wathsara — DISMISSED** [review]

**Review by Wathsara — DISMISSED** [review]

**Review by Wathsara — DISMISSED** [review]

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java [hunk 4]
```diff
@@ -849,11 +854,66 @@ public static String getUserConsentURL(OAuth2Parameters params, String loggedInU
...
+        if (params != null) {
+            sp = OAuth2Util.getServiceProvider(params.getClientId());
+        }
+        return sp;
+    }
+
+    private static boolean isExternalizedConsentPageEnabledForSP(ServiceProvider sp) {
+
+        if (sp == null) {
+            throw new IllegalArgumentException("A null reference received for service provider.");
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java:877]: A better error message would be good.

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java [hunk 5]
```diff
@@ -849,11 +854,66 @@ public static String getUserConsentURL(OAuth2Parameters params, String loggedInU
...
+                    sp.getApplicationName() + " with id: " + sp.getApplicationID());
+        }
+
+        return isEnabled;
+    }
+
+    private static String getExternalConsentUrlForSP(ServiceProvider sp) throws OAuthSystemException {
+
+        if (sp == null) {
+            throw new IllegalArgumentException("A null reference received for service provider.");
```
Comment (mefarazath) [inline components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java:896]: A better error message would be good.

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java (hunk 4 above)
Comment (sandushi) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java]: Resolved by https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/2030/commits/622be8b0e8069d5383ccb968ef7703a90d2c9d44

File: components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java (hunk 5 above)
Comment (sandushi) [reply components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/util/EndpointUtil.java]: Resolved by https://github.com/wso2-extensions/identity-inbound-auth-oauth/pull/2030/commits/622be8b0e8069d5383ccb968ef7703a90d2c9d44

**Review by mefarazath — APPROVED** [review]

**Review by Wathsara — APPROVED** [review]

**Pull request merged**